### PR TITLE
Replace DEPLOYED_VIA_OLM with automatic environment detection

### PR DESCRIPTION
**Pull request by gkrenn**
# Description

Currently we manually configure an environment variable to detect if we run on OLM. This is an unnecessary step and should be replaced by a more automatic detection. 

This PR removes the env var and instead checks if the "olm.operatorNamespace" annotation is set on the operator and webhook pods.

## How can this be tested?

Run code in OLM and check if OLM specific behaviour is happening. 

## Checklist
- [x] Unit tests have been updated/added
- [x] PR is labeled accordingly
- [x] I have read and understood the [contribution guidelines](/CONTRIBUTING.md)

### CODE REVIEW

File: src/cmd/webhook/builder.go [hunk 1]
```diff
@@ -97,7 +97,12 @@ func (builder CommandBuilder) buildRun() func(*cobra.Command, []string) error {
 			return err
 		}
 
-		if !builder.isDeployedViaOlm {
+		isDeployedViaOLM, err := kubesystem.IsDeployedViaOlm(webhookManager.GetClient(), builder.podName, builder.namespace)
```
Comment (meik99) [inline src/cmd/webhook/builder.go:100]: IsDeployedViaOlm could also be wrapped in a function like in the operator builder

File: src/kubesystem/kubesystem.go [hunk 2]
```diff
@@ -20,6 +23,23 @@ func GetUID(clt client.Reader) (types.UID, error) {
...
+	pod := &corev1.Pod{}
+	err := clt.Get(context.TODO(), types.NamespacedName{Name: podName, Namespace: podNamespace}, pod)
+	if err != nil {
+		return false, err
+	}
+	if _, ok := pod.Annotations[olmSpecificAnnotation]; ok {
+		return true, nil
+	} else {
+		return false, nil
+	}
```
Comment (meik99) [inline src/kubesystem/kubesystem.go:36]:
```suggestion
    _, isDeployedViaOlm := pod.Annotations[olmSpecificAnnotation];
    return isDeployedViaOlm, nil
```

File: src/webhook/mutation/pod_mutator/pod_mutator.go [hunk 3]
```diff
@@ -58,6 +59,13 @@ func (webhook *podMutatorWebhook) Handle(ctx context.Context, request admission.
 	if err != nil {
 		return silentErrorResponse(mutationRequest.Pod, err)
 	}
+	// in case of olm deployment, all pods are sent to us
+	// but not all of them need to be mutated,
+	// therefore their namespace might not have a dynakube assigned
+	// in which case we don't need to do anything
+	if mutationRequest == nil {
+		return emptyPatch
+	}
```
Comment (meik99) [inline src/webhook/mutation/pod_mutator/pod_mutator.go:68]:
```suggestion
	if noMutationRequired(mutationRequest) {
	...

func noMutationRequired(mutationRequest) bool {
  return mutationRequest == nil
}
```

the comment wouldn't be needed

File: src/kubesystem/kubesystem.go [hunk 4]
```diff
@@ -20,6 +23,23 @@ func GetUID(clt client.Reader) (types.UID, error) {
...
+}
+
+func CreateDefaultClient() (client.Client, error) {
+	kubeCfg, err := config.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+	return client.New(kubeCfg, client.Options{})
```
Comment (meik99) [inline src/kubesystem/kubesystem.go:44]: error is external, needs to be wrapped
```suggestion
    kubeClient, err := client.New(kubeCfg, client.Options{})
	return kubeClient, errors.WithStack(err)
```

File: src/webhook/mutation/pod_mutator/register.go [hunk 5]
```diff
@@ -49,11 +49,17 @@ func registerInjectEndpoint(mgr manager.Manager, webhookNamespace string, webhoo
 		return errors.WithStack(err)
 	}
 
+	deployedViaOLM, err := kubesystem.IsDeployedViaOlm(apiReader, webhookPodName, webhookNamespace)
+	if err != nil {
+		return errors.WithStack(err)
```
Comment (meik99) [inline src/webhook/mutation/pod_mutator/register.go:54]: Error is internal, must not be wrapped
```suggestion
		return err
```

File: src/webhook/mutation/pod_mutator/request.go [hunk 6]
```diff
@@ -23,9 +23,11 @@ func (webhook *podMutatorWebhook) createMutationRequestBase(ctx context.Context,
 	if err != nil {
 		return nil, err
 	}
-	dynakubeName, err := getDynakubeName(*namespace)
-	if err != nil {
+	dynakubeName, err := getDynakubeName(*namespace, webhook.deployedViaOLM)
+	if err != nil && !webhook.deployedViaOLM {
 		return nil, err
+	} else if err != nil {
+		return nil, nil
```
Comment (meik99) [inline src/webhook/mutation/pod_mutator/request.go:30]: the comment from above about olm would fit better here

File: src/webhook/mutation/pod_mutator/request.go [hunk 7]
```diff
@@ -55,14 +57,10 @@ func getNamespaceFromRequest(ctx context.Context, apiReader client.Reader, req a
...
-func getDynakubeName(namespace corev1.Namespace) (string, error) {
+func getDynakubeName(namespace corev1.Namespace, deployedViaOLM bool) (string, error) {
 	dynakubeName, ok := namespace.Labels[mapper.InstanceLabel]
 	if !ok {
-		var err error
-		if !kubesystem.DeployedViaOLM() {
-			err = fmt.Errorf("no DynaKube instance set for namespace: %s", namespace.Name)
-		}
-		return dynakubeName, err
+		return "", errors.New(fmt.Sprintf("no DynaKube instance set for namespace: %s", namespace.Name))
```
Comment (meik99) [inline src/webhook/mutation/pod_mutator/request.go:63]:
```suggestion
		return "", errors.Errorf("no DynaKube instance set for namespace: %s", namespace.Name)
```

**Review by meik99 — CHANGES_REQUESTED** [review]

File: src/webhook/mutation/pod_mutator/register.go (hunk 5 above)
Comment (gkrenn) [reply src/webhook/mutation/pod_mutator/register.go:54]: True, also removed it for return values above

File: src/kubesystem/kubesystem.go [hunk 8]
```diff
@@ -20,6 +24,21 @@ func GetUID(clt client.Reader) (types.UID, error) {
...
+		return false, errors.WithStack(err)
+	}
+	_, isDeployedViaOlm := pod.Annotations[olmSpecificAnnotation]
+	return isDeployedViaOlm, nil
+}
+
+func CreateDefaultClient() (client.Client, error) {
+	kubeCfg, err := config.GetConfig()
+	if err != nil {
+		return nil, err
```
Comment (0sewa0) [inline src/kubesystem/kubesystem.go:40]:
```suggestion
		return nil, errors.WithStack(err)
```

**Review by 0sewa0 — APPROVED** [review]

**Review by meik99 — APPROVED** [review]

**Pull request merged**